presto-main/src/test/java/io/prestosql/tests/LogTestDurationListener.java

```diff
@@ -13,30 +13,121 @@
 import org.testng.ITestClass;
 import org.testng.ITestResult;
 
+import javax.annotation.concurrent.GuardedBy;
+
+import java.lang.management.ThreadInfo;
+import java.util.Arrays;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
 
 import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
 import static io.airlift.units.Duration.nanosSince;
 import static java.lang.String.format;
+import static java.lang.management.ManagementFactory.getThreadMXBean;
+import static java.util.stream.Collectors.joining;
 
 public class LogTestDurationListener
-        implements IClassListener, IInvokedMethodListener
+        implements IExecutionListener, IClassListener, IInvokedMethodListener
 {
     private static final Logger LOG = Logger.get(LogTestDurationListener.class);
 
     private static final Duration SINGLE_TEST_LOGGING_THRESHOLD = Duration.valueOf("30s");
     private static final Duration CLASS_LOGGING_THRESHOLD = Duration.valueOf("1m");
+    private static final Duration GLOBAL_IDLE_LOGGING_THRESHOLD = Duration.valueOf("5m");
+
+    private final ScheduledExecutorService scheduledExecutorService;
 
     private final Map<String, Long> started = new ConcurrentHashMap<>();
+    private final AtomicLong lastChange = new AtomicLong(System.nanoTime());
+    private final AtomicBoolean hangLogged = new AtomicBoolean();
+    private final AtomicBoolean finished = new AtomicBoolean();
+    @GuardedBy("this")
+    private ScheduledFuture<?> monitorHangTask;
+
+    public LogTestDurationListener()
+    {
+        scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(daemonThreadsNamed("TestHangMonitor"));
+    }
+
+    @Override
+    public synchronized void onExecutionStart()
+    {
+        resetHangMonitor();
+        finished.set(false);
+        if (monitorHangTask == null) {
+            monitorHangTask = scheduledExecutorService.scheduleWithFixedDelay(this::checkForTestHang, 5, 5, TimeUnit.SECONDS);
+        }
+    }
+
+    @Override
+    public synchronized void onExecutionFinish()
+    {
+        resetHangMonitor();
+        finished.set(true);
+        // do not stop hang task so notification of hung threads will fire
+        // Note: since the monitor uses daemon threads it will not prevent JVM shutdown
+    }
+
+    private void checkForTestHang()
+    {
+        if (hangLogged.get()) {
+            return;
+        }
+
+        Duration duration = nanosSince(lastChange.get());
+        if (duration.compareTo(GLOBAL_IDLE_LOGGING_THRESHOLD) < 0) {
+            return;
+        }
+
+        if (!hangLogged.compareAndSet(false, true)) {
+            return;
```

Comment:
Log an error here or `verify`. The way you init the scheduled task, there should one running at a time.

Reply:
The point of this check is to prevent logging over and over again.